src/main/java/com/arextest/common/filter/ContentFilter.java

```diff
@@ -0,0 +1,124 @@
+  private final boolean recordPayloadWithMetric;
+  private static Method method;
+  public ContentFilter(List<MetricListener> metricListeners, boolean recordPayloadWithMetric) {
+    this.metricListeners = metricListeners;
+    this.recordPayloadWithMetric = recordPayloadWithMetric;
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) {
+    // do nothing
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+      throws IOException, ServletException {
+    if (CollectionUtils.isEmpty(metricListeners)) {
+      chain.doFilter(request, response);
+      return;
+    }
+    if (skipMetric(((HttpServletRequest) request).getMethod())) {
+      chain.doFilter(request, response);
+      return;
+    }
+    HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+    chain.doFilter(httpServletRequest, response);
+    logResponseSize(httpServletRequest, response);
+  }
+
+  private void logResponseSize(HttpServletRequest httpServletRequest,
+      ServletResponse servletResponse) {
+    String clientApp = httpServletRequest.getHeader(CLIENT_APP_HEADER);
+    String category = httpServletRequest.getHeader(CATEGORY_TYPE_HEADER);
+    String requestURI = httpServletRequest.getRequestURI();
+    long requestLength = httpServletRequest.getContentLengthLong();
+    long responseLength = getResponseLength(servletResponse);
+    recordPayloadInfo(clientApp, category, requestURI, requestLength, responseLength);
+  }
+
+  private long getResponseLength(ServletResponse servletResponse) {
+    Class<? extends ServletResponse> responseClass = servletResponse.getClass();
+    String responseClassName = responseClass.getName();
+    if ("org.mortbay.jetty.Response".equals(responseClassName)) {
+      // jetty
+      return getResponseLengthByMethod(servletResponse, "getContentCount");
+    } else if ("org.apache.catalina.connector.ResponseFacade".equals(responseClassName)) {
+      //tomcat 7+
+      return getResponseLengthByMethod(servletResponse, "getContentWritten");
+    } else {
+      return 0L;
+    }
+  }
+
+  private static long getResponseLengthByMethod(ServletResponse servletResponse, String methodName) {
+    if (method == null) {
+      try {
+        method = servletResponse.getClass().getDeclaredMethod(methodName);
+      } catch (NoSuchMethodException e) {
+        LOGGER.error("failed to get declared method. {}", e.getMessage(), e);
+        return 0L;
+      }
+    }
+    try {
+      return (long) method.invoke(servletResponse);
+    } catch (IllegalAccessException | InvocationTargetException e) {
+      LOGGER.error("failed to log response size. {}", e.getMessage(), e);
+    }
+    return 0L;
+  }
+
+  private void recordPayloadInfo(String clientApp, String category,
+      String path, long requestLength, long responseLength) {
+    if (responseLength <= 0L || requestLength <= 0L) {
+      return;
+    }
+    Map<String, String> tags = new HashMap<>(4);
```

Comment:
why set map init size 4? what is the use of?

Reply:
In the current scenario, only 4 map data will be stored at most.
It is mainly used as a burying tag.
eg：path clientapp category type